core/Pos/Core/Common/Types.hs

```diff
@@ -303,12 +302,16 @@ maxCoinVal = 45000000000000000
 -- it would fail with 'error' if the 'Word64' exceeds 'maxCoinVal', but now you
 -- have to 'checkCoin' if you care about that.
 mkCoin :: Word64 -> Coin
-mkCoin = Coin
+mkCoin x =
+    let c = UnsafeCoin x
+    in case runPVerify c of
+          Left e   -> error $ "mkCoin: " <> show e
+          Right () -> c
 
-checkCoin :: MonadError Text m => Coin -> m ()
-checkCoin (Coin c)
-    | c <= maxCoinVal = pure ()
-    | otherwise       = throwError $ "Coin: " <> show c <> " is too large"
+instance PVerifiable Coin where
+    pverifyOne (UnsafeCoin c)
+        | c <= maxCoinVal = pass
+        | otherwise       = pverFail $ "Coin: " <> show c <> " is too large"
```

Comment:
Why don't you just define `verifyCoin` top-level? Is there any value in having this `PVerifiable` class? Seems like an overcomplication.

Reply:
It's not. 
* Forces programmer to think about verification in two levels (self + recursive)
* Unifies the standard about how to name these validation functions
* Forbids multiple verifiers for the same type
* Forces intermediate datatypes to be verified as well (not only sub-sub-children, but also children)